### ISSUE TITLE
Sending a partition key makes targeted multi-partition aggregation queries more expensive

### ISSUE THREAD

**Issue by alesk-kontent** [Azure/azure-cosmos-dotnet-v3#4248]
**Describe the bug**

When I run a targeted multi-partition aggregation query against the same data (two levels of depth), the request charge depends on the presence of the partition key. With a partition key that includes the first level of subpartitioning, the request charge is orders of magnitude higher than if I omit the partition key.

**To Reproduce**

I wrote a simple [console application](https://github.com/Azure/azure-cosmos-dotnet-v3/files/13902287/AggregationIssue.zip) to reproduce the issue. Please download the solution, open the `Program.cs` file and specify a connection string to an Azure Cosmos DB for NoSQL account. The application will create a new database with two containers, non-subpartitioned and subpartitioned. For both of them it will upsert the same set of data, run an aggregation query (with and without a partition key) and display important metrics. You can run the application multiple times, it will create resources and upsert data only once.

Each container contains 100 documents (articles) with the following properties:

- Id : string
- CustomerId : Guid
- Label : string
- Title : string

The value of the `Id` property is a random GUID. The value of the `CustomerId` and `Label` properties is always the same. The value of the `Title` property is unique for each document.

Both containers share the same indexing policy; automatic and consistent. However, the partition key is different. The first container (non-subpartitioned) uses a partition key `/customerId`, so there is a single logical partition with 100 documents. The second container (subpartitioned) uses a hierarchical partition key `/customerId, /id`, so there are 100 hierarchical logical partitions with one document.

Then I run a query `SELECT VALUE COUNT(1) FROM c WHERE c.customerId = @customerId AND c.label = @label` against both containers:

- Non-subpartitioned: I send the customer id as the partition key (in the request options).
- Subpartitioned: I send the customer id as the partition key (in the request options).
- Subpartitioned: I do not send a partition key.

I can observe the following results:

Container: NonSubpartitioned
Partition key: ["df165b31-7641-4664-9549-37862ed806ee"]
Request charge: 3,09

Container: Subpartitioned
Partition key: ["df165b31-7641-4664-9549-37862ed806ee"]
Request charge: 6,27

Container: Subpartitioned
Partition key: <none>
Request charge: 3,09

**Expected behavior**

Sending a partition key that includes the first level of subpartitioning does not make aggregation queries more expensive.

**Actual behavior**

With enough data and a partition key that includes the first level of subpartitioning, the request charge of aggregation queries is orders of magnitude higher.

**Environment summary**

SDK Version: 3.37.1
OS Version: Windows 11 Enterprise (10.0.22631 Build 22631)

**Additional context**

It happens in both `Direct` and `Gateway` mode. Also, even if I set the `PopulateIndexMetrics` request option, the response does not contain the metrics.

In previous experiments we generated more data and captured diagnostics from the SDK. See the results from the subpartitioned container; [with](https://github.com/Azure/azure-cosmos-dotnet-v3/files/13940550/aggregate_query_diagnostics_with_pk.json) the first level of subpartitioning and [without](https://github.com/Azure/azure-cosmos-dotnet-v3/files/13940551/aggregate_query_diagnostics_without_pk.json) a partition key.

**Comment by adityasa** [comment]
@alesk-kontent thanks for reporting the issue. I downloaded your app and confirmed your findings.

Comparing the execution of the queries in the subpartitioned collection:
1. For the query that was executed without _any_ partition key in the request, the query engine is able to push both the filter (WHERE) and aggregation (COUNT) to the index fully. The outcome of this is that the query engine is able to calculate the result without loading _any_ documents.
2. For the query that was executed with the partial partition key in the request, the query engine is able to only push the filter (WHERE) to the index but still requires runtime evaluation. This has to do with the facts that
- hierarchical Partition key filter (supplied in the request) is represented differently in the request and effective partition key range (aka epk range) is different from what user supplied in the API
- the index is for the entire physical partition and therefore the epk filter needs to be re-evaluated at runtime
The outcome of this is that the query engine needs to evaluate the partition key filter and then calculate the aggregate (COUNT) by loading each document returned by the index (using query filter). Because query execution contains 1 index consultation and runtime evaluation, RU consumption is higher than case 1 above (where only 2 index consultations are involved).

For the non-subpartitioned collection, query execution is identical to case 1 above where both the filter and the aggregate can be fully pushed to the index and query is executed without loading any documents.

As a mitigation for case 2, we can enable indexing epk range for such case. In this case, it should allow query execution using 3 index consultations (query filter, epk filter and aggregate). Please note that indexing epk range can be expensive for a large number of unique hierarchical partition keys. Generally speaking, in most cases, the RU consumption of epk filter evaluation should be much lower compared to loading the documents, but there may be outlier cases.

I also confirmed your finding regarding index advisor not returning results and are investigating it actively. The latest released sdk (3.38.0) should have this fix and index advice should flow as expected.
However we also found another issue that affects the advice returned by the index advisor in case 2 above (epk filter) and are actively working on the fix.

Hope this helps. Kindly let me know if you have any subsequent questions.

**Comment by leminh98** [comment]
The issue affecting the advice is addressed and fix in this PR https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4303
